File: orchestrate-subagents/scripts/resolve_model_policy.py
#!/usr/bin/env python3
"""Resolve global + project, common + host-specific subagent model policy."""

from __future__ import annotations

import argparse
import json
import os
import sys
from pathlib import Path
from typing import Any, Dict, Iterable, List, Mapping, Optional, Tuple


SCHEMA_VERSION = 1
POLICY_KEYS = {"schema_version", "routes", "task_overrides"}
HOST_KEYS = {"schema_version", "host", "effort_order", "aliases", "profiles", "constraints"}
PROFILE_KEYS = {"model", "effort", "channel", "dispatch"}
CONSTRAINT_KEYS = {"allowed_models", "minimum_effort"}
DISPATCH_PROVENANCE = {"explicit", "inherited-controller", "host-default"}


class PolicyError(ValueError):
    pass


def user_config_dir(
    env: Optional[Mapping[str, str]] = None,
    platform_name: Optional[str] = None,
    home: Optional[Path] = None,
) -> Path:
    values = dict(os.environ if env is None else env)
    explicit = values.get("ORCHESTRATE_SUBAGENTS_CONFIG")
    if explicit:
        return Path(explicit).expanduser()

    platform_value = sys.platform if platform_name is None else platform_name
    home_path = Path.home() if home is None else home
    suffix = Path("agent-skills") / "orchestrate-subagents"
    if platform_value.startswith("win"):
        base = Path(values.get("APPDATA", str(home_path / "AppData" / "Roaming")))
    elif platform_value == "darwin":
        base = home_path / "Library" / "Application Support"
    else:
        base = Path(values.get("XDG_CONFIG_HOME", str(home_path / ".config")))
    return base / suffix


def repository_root(path: Path) -> Path:
    candidate = path.resolve()
    if candidate.is_file():
        candidate = candidate.parent
    for current in (candidate,) + tuple(candidate.parents):
        if (current / ".git").exists():
            return current
    return candidate


def project_config_dir(repo: Path) -> Path:
    return repository_root(repo) / ".agents" / "orchestrate-subagents"


def _object(value: Any, label: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise PolicyError("%s must be a JSON object" % label)
    return dict(value)


def _check_keys(value: Mapping[str, Any], allowed: set, label: str) -> None:
    unknown = sorted(set(value) - allowed)
    if unknown:
        raise PolicyError("%s has unknown keys: %s" % (label, ", ".join(unknown)))


def _load_json(path: Path, kind: str, host: Optional[str] = None) -> Optional[Dict[str, Any]]:
    if not path.is_file():
        return None
    try:
        value = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise PolicyError("cannot read %s: %s" % (path, exc)) from exc
    data = _object(value, str(path))
    _check_keys(data, POLICY_KEYS if kind == "policy" else HOST_KEYS, str(path))
    if data.get("schema_version") != SCHEMA_VERSION:
        raise PolicyError("%s must declare schema_version %s" % (path, SCHEMA_VERSION))
    if kind == "host" and data.get("host") != host:
        raise PolicyError("%s host must be %r" % (path, host))
    return data


def _merge_dict(base: Mapping[str, Any], overlay: Mapping[str, Any]) -> Dict[str, Any]:
    merged = dict(base)
    for key, value in overlay.items():
        if isinstance(value, dict) and isinstance(merged.get(key), dict):
            merged[key] = _merge_dict(merged[key], value)
        else:
            merged[key] = value
    return merged


def _validate_policy(data: Mapping[str, Any], label: str) -> None:
    for field in ("routes", "task_overrides"):
        values = _object(data.get(field, {}), "%s.%s" % (label, field))
        for key, profile in values.items():
            if not isinstance(key, str) or not isinstance(profile, str) or not profile:
                raise PolicyError("%s.%s entries must map strings to profiles" % (label, field))


def _validate_host(data: Mapping[str, Any], label: str) -> None:
    effort_order = data.get("effort_order")
    if effort_order is not None:
        if (
            not isinstance(effort_order, list)
            or any(not isinstance(item, str) or not item for item in effort_order)
            or len(set(effort_order)) != len(effort_order)
        ):
            raise PolicyError("%s.effort_order must be a unique string array" % label)
    aliases = _object(data.get("aliases", {}), "%s.aliases" % label)
    for key, model in aliases.items():
        if not isinstance(key, str) or not isinstance(model, str) or not model:
            raise PolicyError("%s.aliases entries must map strings to model IDs" % label)
    profiles = _object(data.get("profiles", {}), "%s.profiles" % label)
    for name, raw in profiles.items():
        profile = _object(raw, "%s.profiles.%s" % (label, name))
        _check_keys(profile, PROFILE_KEYS, "%s.profiles.%s" % (label, name))
        for key, value in profile.items():
            if not isinstance(value, str) or not value:
                raise PolicyError("%s.profiles.%s.%s must be a string" % (label, name, key))
        if "dispatch" in profile and profile["dispatch"] not in DISPATCH_PROVENANCE:
            raise PolicyError("%s.profiles.%s.dispatch is invalid" % (label, name))
    constraints = _object(data.get("constraints", {}), "%s.constraints" % label)
    _check_keys(constraints, CONSTRAINT_KEYS, "%s.constraints" % label)
    allowed = constraints.get("allowed_models", [])
    if not isinstance(allowed, list) or any(not isinstance(item, str) or not item for item in allowed):
        raise PolicyError("%s.constraints.allowed_models must be a string array" % label)
    minimum = _object(constraints.get("minimum_effort", {}), "%s.constraints.minimum_effort" % label)
    for model, effort in minimum.items():
        if not isinstance(model, str) or not isinstance(effort, str) or not effort:
            raise PolicyError("%s minimum effort for %r is invalid" % (label, model))


def _stricter_effort(first: str, second: str, effort_rank: Mapping[str, int]) -> str:
    if first not in effort_rank or second not in effort_rank:
        raise PolicyError("unknown effort while merging: %r / %r" % (first, second))
    return first if effort_rank[first] >= effort_rank[second] else second


def _merge_constraints(
    global_value: Mapping[str, Any],
    project_value: Mapping[str, Any],
    effort_rank: Mapping[str, int],
) -> Dict[str, Any]:
    result: Dict[str, Any] = {}
    global_allowed = global_value.get("allowed_models")
    project_allowed = project_value.get("allowed_models")
    if global_allowed is not None and project_allowed is not None:
        project_set = set(project_allowed)
        result["allowed_models"] = [model for model in global_allowed if model in project_set]
    elif global_allowed is not None:
        result["allowed_models"] = list(global_allowed)
    elif project_allowed is not None:
        result["allowed_models"] = list(project_allowed)

    minimum = dict(global_value.get("minimum_effort", {}))
    for model, effort in project_value.get("minimum_effort", {}).items():
        minimum[model] = _stricter_effort(minimum[model], effort, effort_rank) if model in minimum else effort
    if minimum:
        result["minimum_effort"] = minimum
    return result


def _config_paths(root: Path, host: str) -> Tuple[Path, Path]:
    return root / "policy.json", root / "hosts" / (host + ".json")


def resolve(args: argparse.Namespace) -> Dict[str, Any]:
    repo = Path(args.repo).resolve()
    global_root = Path(args.global_config_dir).expanduser() if args.global_config_dir else user_config_dir()
    project_root = Path(args.project_config_dir).expanduser() if args.project_config_dir else project_config_dir(repo)
    global_policy_path, global_host_path = _config_paths(global_root, args.host)
    project_policy_path, project_host_path = _config_paths(project_root, args.host)

    sources: List[str] = []
    policy: Dict[str, Any] = {"routes": {}, "task_overrides": {}}
    for path in (global_policy_path, project_policy_path):
        data = _load_json(path, "policy")
        if data is not None:
            _validate_policy(data, str(path))
            policy = _merge_dict(policy, {key: value for key, value in data.items() if key != "schema_version"})
            sources.append(str(path))

    global_host = _load_json(global_host_path, "host", args.host)
    project_host = _load_json(project_host_path, "host", args.host)
    for path, data in ((global_host_path, global_host), (project_host_path, project_host)):
        if data is not None:
            _validate_host(data, str(path))
            sources.append(str(path))

    host_policy = _merge_dict(global_host or {}, project_host or {})
    effort_order = host_policy.get("effort_order")
    if not isinstance(effort_order, list) or not effort_order:
        raise PolicyError("host %r must define effort_order in external config" % args.host)
    effort_rank = {name: index for index, name in enumerate(effort_order)}
    host_policy["constraints"] = _merge_constraints(
        (global_host or {}).get("constraints", {}),
        (project_host or {}).get("constraints", {}),
        effort_rank,
    )

    profile_name = policy.get("task_overrides", {}).get(args.task_type) if args.task_type else None
    if profile_name is None:
        profile_name = policy.get("routes", {}).get(args.role)
    if not profile_name:
        raise PolicyError("no profile route for role=%r task_type=%r" % (args.role, args.task_type))
    raw_profile = host_policy.get("profiles", {}).get(profile_name)
    if not isinstance(raw_profile, dict):
        raise PolicyError("host %r has no profile %r" % (args.host, profile_name))

    aliases = host_policy.get("aliases", {})
    requested_model = args.model or raw_profile.get("model")
    requested_effort = args.effort or raw_profile.get("effort")
    model = aliases.get(requested_model, requested_model)
    if not model or not requested_effort:
        raise PolicyError("profile %r must resolve model and effort" % profile_name)
    if requested_effort not in effort_rank:
        raise PolicyError("effort %r is absent from host effort_order" % requested_effort)

    constraints = host_policy.get("constraints", {})
    allowed = constraints.get("allowed_models")
    if allowed is not None and model not in allowed:
        raise PolicyError("model %r is outside allowed_models" % model)
    minimum = constraints.get("minimum_effort", {}).get(model)
    if minimum:
        if minimum not in effort_rank:
            raise PolicyError("minimum effort %r is absent from host effort_order" % minimum)
        if effort_rank[requested_effort] < effort_rank[minimum]:
            raise PolicyError("effort %r is below %s minimum %r" % (requested_effort, model, minimum))
    if args.available_model and model not in args.available_model:
        raise PolicyError("model %r is not exposed by the current host schema" % model)
    if args.available_effort and requested_effort not in args.available_effort:
        raise PolicyError("effort %r is not exposed by the current host schema" % requested_effort)
    if args.available_channel and raw_profile.get("channel") not in args.available_channel:
        raise PolicyError("channel %r is not exposed by the current host schema" % raw_profile.get("channel"))

    return {
        "schema_version": SCHEMA_VERSION,
        "host": args.host,
        "role": args.role,
        "task_type": args.task_type,
        "profile": profile_name,
        "model": model,
        "effort": requested_effort,
        "channel": raw_profile.get("channel"),
        "dispatch": raw_profile.get("dispatch"),
        "dispatch_provenance": raw_profile.get("dispatch", "host-default"),
        "selection_source": "session-explicit" if args.model or args.effort else "external-policy",
        "config_source": sources or ["skill-default"],
    }


def _parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--host", required=True)
    parser.add_argument("--repo", default=".")
    parser.add_argument("--role", default="worker")
    parser.add_argument("--task-type")
    parser.add_argument("--global-config-dir")
    parser.add_argument("--project-config-dir")
    parser.add_argument("--model")
    parser.add_argument("--effort")
    parser.add_argument("--available-model", action="append", default=[])
    parser.add_argument("--available-effort", action="append", default=[])
    parser.add_argument("--available-channel", action="append", default=[])
    parser.add_argument("--explain", action="store_true")
    return parser


def _explain(result: Mapping[str, Any]) -> str:
    lines = [
        "profile: %s" % result["profile"],
        "model: %s" % result["model"],
        "effort: %s" % result["effort"],
        "channel: %s" % (result.get("channel") or "host-default"),
        "dispatch: %s" % (result.get("dispatch") or "host-default"),
        "provenance: %s" % result["dispatch_provenance"],
        "selection: %s" % result["selection_source"],
        "source:",
    ]
    lines.extend("  - %s" % value for value in result["config_source"])
    return "\n".join(lines)


def main(argv: Optional[Iterable[str]] = None) -> int:
    args = _parser().parse_args(argv)
    try:
        result = resolve(args)
    except PolicyError as exc:
        print("model policy error: %s" % exc, file=sys.stderr)
        return 2
    print(_explain(result) if args.explain else json.dumps(result, ensure_ascii=False, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: orchestrate-subagents/scripts/resolve_model_policy.mjs
#!/usr/bin/env node
// @ts-check

import { existsSync, readFileSync, realpathSync } from 'node:fs';
import { homedir, platform } from 'node:os';
import { dirname, join, resolve as resolvePath } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { randomUUID } from 'node:crypto';
import { parseJsonStrict } from './contract-tool.mjs';

const CONFIG_SCHEMA_VERSION = 2;
const RESULT_SCHEMA_VERSION = 1;
const HOST_KEYS = new Set(['schema_version', 'host', 'effort_order', 'tier_order', 'tiers', 'dynamic_adjustment']);
const TIER_KEYS = new Set(['models', 'effort', 'channel', 'dispatch']);
const EFFORT_KEYS = new Set(['default', 'min', 'max']);
const DYNAMIC_KEYS = new Set(['enabled', 'max_attempts', 'allowed_actions']);
const DISPATCH_PROVENANCE = new Set(['explicit', 'inherited-controller', 'host-default']);
export const ADJUSTMENT_ACTIONS = new Set(['retry_same', 'raise_effort', 'switch_model', 'promote_tier', 'fresh_context', 'change_strategy']);
export const FAILURE_KINDS = new Set(['implementation_defect', 'reasoning_gap', 'context_gap', 'strategy_gap', 'environment_fault', 'contract_gap', 'safety', 'undecidable']);
const REROUTABLE_FAILURES = new Set(['implementation_defect', 'reasoning_gap', 'context_gap', 'strategy_gap']);
const DIGEST_PATTERN = /^sha256:[0-9a-f]{64}$/u;

export class PolicyError extends Error {}

function expandUserPath(value, home = homedir()) {
  if (value === '~') return home;
  if (value.startsWith('~/') || value.startsWith('~\\')) return join(home, value.slice(2));
  return value;
}

export function resolveUserPath(value, home = homedir()) {
  return resolvePath(expandUserPath(value, home));
}

export function userConfigDir(env = process.env, platformName = platform(), home = homedir()) {
  const explicit = env.ORCHESTRATE_SUBAGENTS_CONFIG;
  if (explicit) return resolveUserPath(explicit, home);
  const suffix = join('agent-skills', 'orchestrate-subagents');
  if (platformName.startsWith('win')) return join(env.APPDATA || join(home, 'AppData', 'Roaming'), suffix);
  if (platformName === 'darwin') return join(home, 'Library', 'Application Support', suffix);
  return join(env.XDG_CONFIG_HOME || join(home, '.config'), suffix);
}

// Capability cache still has a project-scoped facts directory. Model preferences do not load from it.
export function repositoryRoot(candidatePath) {
  const requested = resolveUserPath(candidatePath);
  let current = requested;
  while (!existsSync(current)) {
    const parent = dirname(current);
    if (parent === current) return requested;
    current = parent;
  }
  while (true) {
    if (existsSync(join(current, '.git'))) return current;
    const parent = dirname(current);
    if (parent === current) return requested;
    current = parent;
  }
}

export function projectConfigDir(repoPath) {
  return join(repositoryRoot(repoPath), '.agents', 'orchestrate-subagents');
}

function assertObject(value, label) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) throw new PolicyError(`${label} must be a JSON object`);
  return value;
}

function checkKeys(value, allowed, label) {
  const unknown = Object.keys(value).filter((key) => !allowed.has(key)).sort();
  if (unknown.length) throw new PolicyError(`${label} has unknown keys: ${unknown.join(', ')}`);
}

function uniqueStrings(value, label, { nonEmpty = true } = {}) {
  if (!Array.isArray(value) || (nonEmpty && value.length === 0) || value.some((item) => typeof item !== 'string' || !item.trim()) || new Set(value).size !== value.length) {
    throw new PolicyError(`${label} must be a${nonEmpty ? ' non-empty' : ''} unique string array`);
  }
  return value;
}

function readJson(path, label = path) {
  try { return parseJsonStrict(readFileSync(path, 'utf8')); }
  catch (error) { throw new PolicyError(`cannot read ${label}: ${error instanceof Error ? error.message : String(error)}`); }
}

function loadHostConfig(path, expectedHost) {
  if (!existsSync(path)) return null;
  const config = assertObject(readJson(path), path);
  checkKeys(config, HOST_KEYS, path);
  if (config.schema_version !== CONFIG_SCHEMA_VERSION) throw new PolicyError(`${path} must declare schema_version ${CONFIG_SCHEMA_VERSION}`);
  if (config.host !== expectedHost) throw new PolicyError(`${path} host must be ${JSON.stringify(expectedHost)}`);

  const effortOrder = uniqueStrings(config.effort_order, `${path}.effort_order`);
  const effortRank = new Map(effortOrder.map((name, index) => [name, index]));
  const tierOrder = uniqueStrings(config.tier_order, `${path}.tier_order`);
  const tiers = assertObject(config.tiers, `${path}.tiers`);
  if (Object.keys(tiers).length !== tierOrder.length || tierOrder.some((name) => !Object.hasOwn(tiers, name)) || Object.keys(tiers).some((name) => !tierOrder.includes(name))) {
    throw new PolicyError(`${path}.tier_order and tiers must name the same tiers exactly once`);
  }

  for (const [name, raw] of Object.entries(tiers)) {
    const tier = assertObject(raw, `${path}.tiers.${name}`);
    checkKeys(tier, TIER_KEYS, `${path}.tiers.${name}`);
    uniqueStrings(tier.models, `${path}.tiers.${name}.models`);
    const effort = assertObject(tier.effort, `${path}.tiers.${name}.effort`);
    checkKeys(effort, EFFORT_KEYS, `${path}.tiers.${name}.effort`);
    for (const key of EFFORT_KEYS) {
      if (typeof effort[key] !== 'string' || !effortRank.has(effort[key])) throw new PolicyError(`${path}.tiers.${name}.effort.${key} must appear in effort_order`);
    }
    if (effortRank.get(effort.min) > effortRank.get(effort.default) || effortRank.get(effort.default) > effortRank.get(effort.max)) throw new PolicyError(`${path}.tiers.${name}.effort must satisfy min <= default <= max`);
    if (typeof tier.channel !== 'string' || !tier.channel.trim()) throw new PolicyError(`${path}.tiers.${name}.channel must be a non-empty string`);
    if (!DISPATCH_PROVENANCE.has(tier.dispatch)) throw new PolicyError(`${path}.tiers.${name}.dispatch is invalid`);
  }

  const dynamic = assertObject(config.dynamic_adjustment, `${path}.dynamic_adjustment`);
  checkKeys(dynamic, DYNAMIC_KEYS, `${path}.dynamic_adjustment`);
  if (typeof dynamic.enabled !== 'boolean') throw new PolicyError(`${path}.dynamic_adjustment.enabled must be boolean`);
  if (!Number.isSafeInteger(dynamic.max_attempts) || dynamic.max_attempts < 1) throw new PolicyError(`${path}.dynamic_adjustment.max_attempts must be a positive safe integer`);
  uniqueStrings(dynamic.allowed_actions, `${path}.dynamic_adjustment.allowed_actions`, { nonEmpty: false });
  const unknownActions = dynamic.allowed_actions.filter((action) => !ADJUSTMENT_ACTIONS.has(action));
  if (unknownActions.length) throw new PolicyError(`${path}.dynamic_adjustment.allowed_actions has invalid actions: ${unknownActions.join(', ')}`);
  return config;
}

function rankOf(order, value, label) {
  const rank = order.indexOf(value);
  if (rank < 0) throw new PolicyError(`${label} ${JSON.stringify(value)} is not configured`);
  return rank;
}

function chooseModel(tier, explicitModel, availableModels, userExplicit) {
  if (explicitModel) {
    if (!userExplicit && !tier.models.includes(explicitModel)) throw new PolicyError(`model ${JSON.stringify(explicitModel)} is outside the selected local tier`);
    if (availableModels.length && !availableModels.includes(explicitModel)) throw new PolicyError(`model ${JSON.stringify(explicitModel)} is not exposed by the current host schema`);
    return explicitModel;
  }
  if (!availableModels.length) throw new PolicyError('model discovery is unavailable; persisted tier candidates cannot be validated');
  const selected = tier.models.find((model) => availableModels.includes(model));
  if (!selected) throw new PolicyError('none of the selected tier models is exposed by the current host schema');
  return selected;
}

function validateEffort(tier, effortOrder, effort, availableEfforts) {
  const effortRank = new Map(effortOrder.map((name, index) => [name, index]));
  if (!effortRank.has(effort)) throw new PolicyError(`effort ${JSON.stringify(effort)} is absent from effort_order`);
  if (effortRank.get(effort) < effortRank.get(tier.effort.min) || effortRank.get(effort) > effortRank.get(tier.effort.max)) throw new PolicyError(`effort ${JSON.stringify(effort)} is outside tier range ${JSON.stringify(tier.effort.min)}..${JSON.stringify(tier.effort.max)}`);
  if (availableEfforts.length && !availableEfforts.includes(effort)) throw new PolicyError(`effort ${JSON.stringify(effort)} is not exposed by the current host schema`);
}

function validateAdjustment(args, config, selection, previous) {
  if (args.attempt === 1) {
    if (previous || args.action || args.failure_kind || args.failure_ref) throw new PolicyError('initial dispatch cannot declare retry inputs');
    return { action: 'initial', failure_kind: null, failure_ref: null, previous_attempt_id: null };
  }
  if (!config.dynamic_adjustment.enabled) throw new PolicyError('dynamic adjustment is disabled by local config');
  if (args.attempt > config.dynamic_adjustment.max_attempts) throw new PolicyError(`attempt ${args.attempt} exceeds local max_attempts ${config.dynamic_adjustment.max_attempts}`);
  if (!previous) throw new PolicyError('retry dispatch requires --previous-dispatch');
  const previousPatch = previous.dispatch_record_patch;
  if (previous.schema_version !== RESULT_SCHEMA_VERSION || previous.record_type !== 'model-policy-resolution' || previous.host !== args.host || !previousPatch || typeof previousPatch !== 'object' || Array.isArray(previousPatch)) throw new PolicyError('previous resolution schema/host mismatch');
  if (previousPatch.attempt !== args.attempt - 1) throw new PolicyError('previous resolution attempt is not immediately prior');
  if (!ADJUSTMENT_ACTIONS.has(args.action)) throw new PolicyError('retry dispatch requires a valid --action');
  if (!config.dynamic_adjustment.allowed_actions.includes(args.action)) throw new PolicyError(`adjustment action ${JSON.stringify(args.action)} is outside local config`);
  if (!FAILURE_KINDS.has(args.failure_kind)) throw new PolicyError('retry dispatch requires a valid --failure-kind');
  if (!REROUTABLE_FAILURES.has(args.failure_kind)) throw new PolicyError(`failure kind ${JSON.stringify(args.failure_kind)} must stop/re-contract instead of model rerouting`);
  if (!DIGEST_PATTERN.test(args.failure_ref || '')) throw new PolicyError('retry dispatch requires a sha256 --failure-ref');

  const sameTier = previousPatch.tier === selection.tier;
  const sameModel = previousPatch.model === selection.model;
  const sameEffort = previousPatch.reasoning_effort === selection.effort;
  const effortRank = new Map(config.effort_order.map((name, index) => [name, index]));
  if (args.action === 'retry_same' && !(sameTier && sameModel && sameEffort)) throw new PolicyError('retry_same must preserve tier, model and effort');
  if (args.action === 'raise_effort' && !(sameTier && sameModel && effortRank.get(selection.effort) > effortRank.get(previousPatch.reasoning_effort))) throw new PolicyError('raise_effort must keep tier/model and increase effort');
  if (args.action === 'switch_model' && sameModel) throw new PolicyError('switch_model must change the model');
  if (args.action === 'promote_tier' && rankOf(config.tier_order, selection.tier, 'tier') <= rankOf(config.tier_order, previousPatch.tier, 'previous tier')) throw new PolicyError('promote_tier must move to a higher configured tier');
  if (args.action === 'fresh_context' && !(sameTier && sameModel && sameEffort)) throw new PolicyError('fresh_context changes worker context, not model parameters');
  return { action: args.action, failure_kind: args.failure_kind, failure_ref: args.failure_ref, previous_attempt_id: previousPatch.attempt_id };
}

export function resolve(args) {
  const configRoot = args.config_dir ? resolveUserPath(args.config_dir) : userConfigDir();
  const configPath = join(configRoot, 'hosts', `${args.host}.json`);
  const config = loadHostConfig(configPath, args.host);
  if (!config) throw new PolicyError(`no local host config: ${configPath}`);
  const tierName = args.tier;
  const tier = Object.hasOwn(config.tiers, tierName) ? config.tiers[tierName] : null;
  if (!tier) throw new PolicyError(`tier ${JSON.stringify(tierName)} is not configured`);
  const availableModels = args.available_model || [];
  const availableEfforts = args.available_effort || [];
  const availableChannels = args.available_channel || [];
  const userExplicit = args.user_explicit === true;
  if (userExplicit && (!args.model || !args.effort)) throw new PolicyError('--user-explicit requires explicit --model and --effort values');
  const model = chooseModel(tier, args.model, availableModels, userExplicit);
  const effort = args.effort || tier.effort.default;
  validateEffort(tier, config.effort_order, effort, availableEfforts);
  if (availableChannels.length && !availableChannels.includes(tier.channel)) throw new PolicyError(`channel ${JSON.stringify(tier.channel)} is not exposed by the current host schema`);
  if ((!availableModels.length || !availableEfforts.length) && !userExplicit) throw new PolicyError('persisted local config requires live model and effort discovery');

  const previous = args.previous_dispatch ? assertObject(readJson(resolveUserPath(args.previous_dispatch), args.previous_dispatch), args.previous_dispatch) : null;
  const selection = { tier: tierName, model, effort };
  const adjustment = validateAdjustment(args, config, selection, previous);
  const dispatchRecordPatch = {
    attempt_id: args.attempt_id || randomUUID(),
    attempt: args.attempt,
    previous_attempt_id: adjustment.previous_attempt_id,
    tier: tierName,
    model,
    reasoning_effort: effort,
    adjustment_action: adjustment.action,
    failure_kind: adjustment.failure_kind,
    failure_ref: adjustment.failure_ref,
    selection_reason: args.selection_reason,
    config_source: [configPath],
    configuration_state: userExplicit ? 'user-explicit' : 'persisted-config',
    model_resolution_state: availableModels.length ? 'discovered-and-validated' : 'user-explicit-unverifiable',
    dispatch_provenance: tier.dispatch,
    max_attempts: config.dynamic_adjustment.max_attempts,
  };
  return {
    schema_version: RESULT_SCHEMA_VERSION,
    record_type: 'model-policy-resolution',
    host: args.host,
    tier_rank: rankOf(config.tier_order, tierName, 'tier'),
    channel: tier.channel,
    selection_source: userExplicit ? 'user-explicit' : 'local-config',
    dispatch_record_patch: dispatchRecordPatch,
  };
}

export function explain(result) {
  const patch = result.dispatch_record_patch;
  return [
    `attempt: ${patch.attempt}`,
    `tier: ${patch.tier}`,
    `model: ${patch.model}`,
    `effort: ${patch.reasoning_effort}`,
    `action: ${patch.adjustment_action}`,
    `channel: ${result.channel}`,
    `provenance: ${patch.dispatch_provenance}`,
    `configuration_state: ${patch.configuration_state}`,
    `model_resolution_state: ${patch.model_resolution_state}`,
    `selection_reason: ${patch.selection_reason}`,
    'source:',
    ...patch.config_source.map((value) => `  - ${value}`),
  ].join('\n');
}

export function parseCli(argv = process.argv.slice(2)) {
  const options = {
    host: '', config_dir: null, tier: '', model: null, effort: null, selection_reason: '', user_explicit: false,
    attempt: 1, attempt_id: null, previous_dispatch: null, action: null, failure_kind: null, failure_ref: null,
    available_model: [], available_effort: [], available_channel: [], explain: false,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const takeValue = () => {
      const value = argv[i + 1];
      if (value === undefined || value.startsWith('--')) throw new PolicyError(`${arg} 缺少值`);
      i += 1;
      return value;
    };
    if (arg === '--host') options.host = takeValue();
    else if (arg === '--config-dir') options.config_dir = takeValue();
    else if (arg === '--tier') options.tier = takeValue();
    else if (arg === '--model') options.model = takeValue();
    else if (arg === '--effort') options.effort = takeValue();
    else if (arg === '--selection-reason') options.selection_reason = takeValue();
    else if (arg === '--user-explicit') options.user_explicit = true;
    else if (arg === '--attempt') options.attempt = Number(takeValue());
    else if (arg === '--attempt-id') options.attempt_id = takeValue();
    else if (arg === '--previous-dispatch') options.previous_dispatch = takeValue();
    else if (arg === '--action') options.action = takeValue();
    else if (arg === '--failure-kind') options.failure_kind = takeValue();
    else if (arg === '--failure-ref') options.failure_ref = takeValue();
    else if (arg === '--available-model') options.available_model.push(takeValue());
    else if (arg === '--available-effort') options.available_effort.push(takeValue());
    else if (arg === '--available-channel') options.available_channel.push(takeValue());
    else if (arg === '--explain') options.explain = true;
    else throw new PolicyError(`unknown option: ${arg}`);
  }
  if (!options.host) throw new PolicyError('缺少 --host');
  if (!options.tier) throw new PolicyError('缺少 --tier');
  if (!options.selection_reason || /[\r\n]/u.test(options.selection_reason)) throw new PolicyError('缺少合法 --selection-reason');
  if (!Number.isSafeInteger(options.attempt) || options.attempt < 1) throw new PolicyError('--attempt 必须为正安全整数');
  return options;
}

export function main(argv = process.argv.slice(2)) {
  const options = parseCli(argv);
  const result = resolve(options);
  process.stdout.write(options.explain ? `${explain(result)}\n` : `${JSON.stringify(result, null, 2)}\n`);
  return 0;
}

function entry() {
  try { return realpathSync(process.argv[1]) === realpathSync(fileURLToPath(import.meta.url)); }
  catch { return pathToFileURL(resolvePath(process.argv[1] ?? '')).href === import.meta.url; }
}

if (entry()) {
  try { main(); }
  catch (error) {
    if (error instanceof PolicyError) {
      process.stderr.write(`model policy error: ${error.message}\n`);
      process.exit(2);
    }
    throw error;
  }
}
